air/src/proof/mod.rs

```diff
@@ -198,3 +199,54 @@ fn get_conjectured_security(
         collision_resistance,
     )
 }
+
+/// Computes proven security level for the specified proof parameters.
+fn get_proven_security(
+    options: &ProofOptions,
+    base_field_bits: u32,
+    lde_domain_size: u64,
+    collision_resistance: u32,
+) -> u32 {
+    let extension_field_bits = (base_field_bits * options.field_extension().degree()) as f64;
+    let blowup_bits = log2(options.blowup_factor()) as f64;
+    let num_fri_queries = options.num_queries() as f64;
+    let lde_size_bits = lde_domain_size.trailing_zeros() as f64;
+
+    // m is a parameter greater or equal to 3.
+    // A larger m gives a worse field security bound but a better query security bound.
+    // An optimal value of m is then a value that would balance field and query security
+    // but there is no simple closed form solution.
+    // This sets m so that field security is equal to the best query security for any value
+    // of m, unless the calculated value is less than 3 in which case it gets rounded up to 3.
+    let mut m = extension_field_bits;
+    m -= options.grinding_factor() as f64;
+    m -= (num_fri_queries + 3.0) / 2.0 * blowup_bits;
+    m -= 2.0 * lde_size_bits;
+    m /= 7.0;
+    m = 2.0_f64.powf(m);
+    m -= 0.5;
+    m = m.max(3.0);
+
+    // compute field security
+    // this considers only the third component given in the corresponding part of eq. 20
+    // in https://eprint.iacr.org/2021/582, i.e. (m+1/2)^7.n^2 / (2\rho^1.5.q) as all
+    // other terms are negligible in comparison.
+    let field_security = (extension_field_bits + 1.0
+        - 3.0 / 2.0 * blowup_bits
```

Comment:
```suggestion
        + 3.0 / 2.0 * blowup_bits
```
There's actually a mistake in the sign here. I'll fix it.

Reply:
Nevermind, double-checking this it is correct.

Reply:
Hmm - I think there is still something odd here. It seems like increasing grinding factor lowers proven security. For example:

Running `lamport-a` example with `-q 27 -b 16 -g 0 `, I get:
```
Proof security: 107 bits (52 proven)
```

But if I run it with `q 27 -b 16 -g 20`, I get:
```
Proof security: 108 bits (46 proven)
```

Reply:
Oh interesting, I was running tests without grinding to match Ulrich's paper values. Let me have a deeper look.

Reply:
Ah, this arises because the dynamic update of $m$ _assumes_ we're considering the grinding factor, while the contribution threshold is 80 bits, which we very rarely reach with parameters targeted for the conjectured security. The easiest solution seems to fully remove the conditional addition (only for the proven security estimate), as otherwise we may run into circular dependencies for calculating $m$. What do you think?

Reply:
> The easiest solution seems to fully remove the conditional addition

This would basically remove the floor, right? I think this is fine but we can probably just compute security level twice  - once assuming there is grinding factor and once assuming there isn't - and then if the value without grinding factor is over some threshold, we could pick the the result which includes the grinding factor.

Having said this, there might be two things to consider:
- Maybe the floor for proven security should be lower (i.g., 60 bits).
- Maybe we don't need the floor at all?

I can see arguments for both. Probably more for the former and less for the latter.

Reply:
Although, we may be have sometimes the case where the pre_query security level is lower than the query security level, in which case the suggested display above may look weird (with something like `k bits proven, including n bits of grinding`, where `n > k`).

Reply:
> The threshold is here to ensure that we don't rely solely on a huge PoW to reach sufficient security, and keeping one would mean we would almost never consider the impact on grinding.

Agreed. Let's just remove the floor for proven security. I guess it will be rather unlikely (or maybe even impossible?) for proven security (without floor) to be higher than conjectured security (with the floor).